### ISSUE TITLE
Wrong structure when matching protocols

### ISSUE THREAD

**Issue by diogomonica** [packetfu/packetfu#71]
The way the code is currently organized right now, packetfu tries a top-down matching approach for protocols:

`@packet_classes_by_layer = @packet_classes.sort_by { |pclass| pclass.layer }.reverse`

This means that it starts by trying to see if the packet is a PacketFu::HSRPPacket valid packet and goes through all of them until it reaches PacketFu::InvalidPacket:

`p = classes.detect { |pclass| pclass.can_parse?(packet) }.new`

I believe this is the wrong approach. Protocols should always be validated from the outside layers (lower layer protocols) to the inner layers (higher protocol layers). The outer protocols actually provide fields to let the upper layers know what inner protocols is being used. For example, the Ethernet frame provides the value 0x0800 to indicate an IP packet on top of it, that in turn provides the value 0x06 to indicate a TCP segment on top.

The way this seems like it should operate is:
- We list all packet types by layers (lower layers to upper layers).
- We do a .can_parse? of all up-to layer 3 protocols
- We start following the "Type" (Ethernet) and the "Protocol" numbers to parse the packet above those

I'm having problems when implementing 802.11 decoding (Layer 2) because if we do it as it currently works, this is what happens on the code path to a random IP Packet:
- Can_parse? returns false for all protocols until the IP Packet
- Test for IP Packet can_parse?
- Assumes explicitly that the underlying packet is an Ethernet packet (checks for EthPacket.can_parse?)
- If it can parse ethernet, returns true; otherwise, returns false

This means that to allow for another type besides EthPacket, we have to put code inside of the IP packet class (ip.rb) that is going to try to decide what the underlying frame is (by guessing). It would have been a lot better if instead of each individual layer 3 protocol individually testing all the lower level protocols, if we had the code start by testing the layer 2 ones and then just walking up via the Types.

Even if we don't want to walk up the protocols by using type fields (and we want to guess everything by testing all the classes as we do right now), we could also just do it bottom-to-top and save some code ugliness.

**Comment by todb-r7** [comment]
Yep, this is a structural issue that I struggle with for non-Eth packets, and for IPv6 stuff.

One contributor, on #68, short circuited a lot of the top-down-ness by just tossing an ip6 chooser option on creation. I'm happy to talk about this more on this issue on parsing.
